### DISCUSSION TITLE
How do I pass data from an edge middleware to whatever function might run next after a rewrite?

### DISCUSSION THREAD

**Discussion by AaronBeaudoin** [vercel/vercel#8487] [Help]
_This discussion is a refinement of https://github.com/vercel/vercel/discussions/8423, which was a much earlier attempt to narrow down the specific problem I'm facing._

## The Problem

When rewriting or redirecting using edge middleware, any "next function" that is run after the middleware does not seem to get any of the data from the middleware. Query parameters are not passed along from the edge middleware, and any headers set are also not available. (However, both query parameters and headers are still available eventually in the browser when rewriting.)

## Minimal Reproduction

Please take a look at my minimal reproduction [here](https://github.com/AaronBeaudoin/vercel-edge-middleware-test). See the README for some more detailed information, and visit https://vercel-edge-middleware-test.vercel.app to play around with it yourself. Solving this problem is critical to going live with a production project I'm working on for my company, so any help is much appreciated!

**Comment by AaronBeaudoin** [comment]
**Update:** Looks like I'm far from the only person who wants to do this. [Here](https://github.com/vercel/next.js/discussions/31188) is a highly upvoted discussion full of Next.js users under the impression that this is a limitation of the framework, when really the problem is Vercel's edge middleware.

**Reply by leerob** [reply]
This is not accurate. This is a Next.js issue mentioned in the linked thread and unrelated to Vercel's infrastructure.

**Comment by amyegan** [comment]
Thanks for starting a new discussion for clarity on the current state of the problem you're seeing. The Engineering team is looking into possible causes

**Reply by AaronBeaudoin** [reply]
Hi @amyegan, just checking in to see if there's been any progress on this. I'm sort of at a standstill with our company project until we get a solution for this, unless I change gears and dig into alternatives like Cloudflare Workers.

**Reply by AaronBeaudoin** [reply]
**Update**: I've spent the whole day digging into Cloudflare [Workers](https://developers.cloudflare.com/workers) and [Workers KV](https://developers.cloudflare.com/workers/learning/how-kv-works) documentation and examples, setting up an account and giving them a try, and so far it seems like they're just a more flexible/configurable/intuitive version of Vercel's edge functions, prerender functions, and edge middleware. To anyone reading this—Cloudflare might be worth a look!

Everything that I've been trying to get working on Vercel with frustratingly infrequent and sparse support so far seems easily achievable with Cloudflare Workers and Workers KV—at a much better price point as well! I may have yet to run into some walls, but so far, so good! **Vercel, you've seriously got to step it up here**:

- Not everyone uses Next.js. Please, stop putting everyone else in second-class land!
- Don't advertise edge middleware as "production ready" when you can't even modify incoming responses (a _basic_ use case).
- People who go over the $20/month Pro limits probably aren't ready to jump right up to a $5,000+/month enterprise plan. You _really need_ to either have something reasonable in-between, or making the overage cost scaling more linear.
- The symlink method for providing initial static "fallback" versions of prerender function is the most unintuitive way I could possibly imagine for achieving the end goal of just giving the function an initial output at build time.
- Don't force me to use the `/api` directory for my functions! To be fair, what I'm trying to do is complicated enough that I was always going to end up using the Build Output API directly anyways, but still—how is there not a config option to manually set the directory to look for functions in??

All these things can be fixed! Just communicate with your community. I'm trying to get a project live, yet at every wall I feel like my success with Vercel is the very last thing on Vercel's priority list. I had 3 calls with Netlify a few weeks ago, and I could tell they really wanted our business. Such has not been the case _at all_ with Vercel. Excuse my rant.

**Reply by amyegan** [reply]
I'm afraid we have a good news + bad news situation 😞 

_The (sort of) good news:_ The Engineering team has identified a discrepancy between rewriting from Edge Middleware to a [Lambda](https://vercel.com/docs/concepts/functions/serverless-functions) vs. an [Edge function](https://vercel.com/docs/concepts/functions/edge-functions) (currently in [beta](https://vercel.com/docs/concepts/release-phases#beta)). It looks like you might be able to get this working with Lambda, but it isn't going to work with Edge Functions until the Engineers can make a change on our end.

_The bad news:_ That change isn't likely to come fast enough for your project since it sounds like you're bumping up against a deadline. I don't have an ETA for an Edge Functions change to be complete.

If Serverless Functions (Lambda) would work for your project, then I would recommend giving that approach a try. I will share updates here as I get more concrete news from the Engineering team. (I know that's too late to be helpful for Aaron, but should help others in the future).

**Reply by TooTallNate** [reply]
Regarding the query string parameter passing, note that it is possible to forward query parameters to the target resource, but the query parameter needs to include a value (after an equal sign). The current behavior is that query parameters with empty values gets stripped away before invoking the target function. Ex:

* https://vercel-edge-middleware-test.vercel.app/rewrite?test=1 Works, query parameter is preserved / forwarded
* https://vercel-edge-middleware-test.vercel.app/rewrite?test= Query parameter is removed
* https://vercel-edge-middleware-test.vercel.app/rewrite?test Query parameter is removed

I do agree that this is peculiar behavior and we're discussing internally the feasibility of changing that behavior to be less… surprising. But maybe that explanation will help get you unblocked for the time being. Cheers!

**Reply by AaronBeaudoin** [reply]
That behavior is indeed very confusing, but additionally I'd say it's sort of "standard breaking" too. Passing a query parameter without a value is not undefined behavior; everywhere else I've seen, it is treated as essentially passing an empty string for that parameter, but the parameter **still exists**.

 I was able to unblock my project, but I did so by switching to Cloudflare Workers. So far it is much, much easier and simpler to use. Vercel's Edge Functions and Edge Middleware are super confusing; I'm trying to understand how you guys are expecting to add value on top of Cloudflare Workers by making them _less easy to work with_.

**Reply by AaronBeaudoin** [reply]
The particular use case for me on Vercel was to use prerender functions to cache my data and then manually revalidate the data later from the CMS system itself. However, due the weird "symlink" method for giving specific prerender function queries an initial value, I needed to use another function as a "dispatcher" to handle forwarding requests to my prerender functions to keep things simple in the application. Basically when making a request like `/data/product?name=Chair`, the "dispatcher" function would be responsible for checking if that particular route had been generated already at build time (by providing a prerender "fallback") and then sending the request to that symlinked prerender function. The dispatcher function also needs to be fast of course, otherwise I'm adding a lot of overhead to every request, so in concept the best solution is to make it simply an edge middleware, since I should be able to simply use an edge middleware for dispatching routes like that and add close to zero overhead.

Problem is, this is all super over-engineered because it's trying to solve problems created by Vercel.
- If I could just provide a prerender function with multiple "fallbacks", and specify which one to show for a particular route, that would make my life 10x easier. But then, in order to do that, I'd need to be able to call my prerender function from more than just single path, so one limitation leads to another. If this wasn't a problem, I wouldn't even need a "dispatcher" function at all and the whole architecture would suddenly be much simpler.
- If I could just pass on data easily from my edge middleware "dispatcher" to the prerender function, that would also make my life 10x easier. Of course, any robust middleware makes doing something like this really simple, because it's such a common use case.

Guys, you need to be aware that Cloudflare Workers made all of this ridiculously easy for me:
- Instead of trying to shoehorn prerender functions to do what I'm trying to accomplish, the Worker KV store does exactly the same thing but with 10x less work. I just check for my route as a key in the store, and if it exists, return the data, otherwise run my "function" to generate the data so next time it'll be in the "cache" (the KV store). That's how it should be. EASY.
- To "revalidate" my key in the store, I just call some `/refresh` route (or whatever I want to call it) and add my own authentication to it and use some query parameters or something to tell myself which key to replace with fresh data.

In practice this is the exact same thing I was trying to do with Vercel, but with no walls in my way. Take notes.

**Comment by steven-tey** [comment]
Hey @AaronBeaudoin! One workaround for this is you could do that by encoding the data in a hash – we discussed this exact method in the [Builder.io + Vercel webinar](https://youtu.be/whn2o0LlJyA?t=1431) that I hosted the other day. I know it's not ideal but if you wanna do a full static build (with `getStaticProps`) this is your best bet. Otherwise if you're using `getServerSideProps` you could also encode the data in query params and pass it to the page that way! :)

**Reply by AaronBeaudoin** [reply]
Thanks steven-tey, but it sounds like you're assuming I'm using Next.js. I'm not.

**Reply by leerob** [reply]
What framework are you using?

**Reply by AaronBeaudoin** [reply]
I'm using [Vite](https://vitejs.dev) + [`vite-plugin-ssr`](https://vite-plugin-ssr.com). To be honest, though, the insane struggle to get edge and prerender functions working how I want on Vercel unfortunately eventually just forced me to switch to other platforms. It's a shame, because I was really pushing for Vercel.

- For edge functions, [Cloudflare Workers](https://workers.cloudflare.com/) is (in my opinion) unrivaled.
- For serverless functions, Vercel is viable. Most platforms have a solid offering in this area. I'd actually still pass on Vercel now even in this space due to Vercel's "jumpscare" pricing (1,000 GB-hours = $20, but 2,000 GB-hours = $420). The alternative I'd recommend is [Netlify](https://www.netlify.com/products/#netlify-functions) due to better DX and flexibility (for example, you can actually choose a different directory other than `api/` to put your functions in, a basic config need). But Cloudflare Workers is so good that you might not even need serverless functions for a lot of projects. I don't for mine, and it's a full e-commerce site.

For clarity, these are the specific issues that, if solved, would make Vercel the king for me:

- Linear pricing. If I use functions ten times as much, I should get charged precisely ten times as much.
- A "framework-less" platform design philosophy. Everything seems to revolve around Next.js right now.
- Allow prerender functions to handle multiple routes/paths, and set initial outputs per-route.
- Allow customizing the directories for serverless functions and any other current/future Vercel features.
- An easy, predictable way to pass data from edge middleware onward to whatever will be called afterward.
